### PR TITLE
feat(component): add tooltip 

### PR DESCRIPTION
**Pull request by igorwessel**
## Infos

[Task](https://juntossomosmais.monday.com/boards/1521247929/pulses/5409625274)

Closes https://github.com/juntossomosmais/atomium/issues/113

#### What is being delivered?

- Tooltip component

#### What impacts?

Nothing because is a new component.

#### Reversal plan

Revert this merge

#### Evidences

https://github.com/juntossomosmais/atomium/assets/54173994/285866a3-bec5-49c0-9250-8cb843dcb27d

### CODE REVIEW

File: packages/core/src/components/tooltip/tooltip.spec.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+describe('AtomTooltip', () => {
+  const createTooltip = (id: string) => `
+      <button id="${id}" aria-describedby="${id}--tooltip">Hover</button>
+      <atom-tooltip id="${id}--tooltip" element="${id}" action="hover">John Doe</atom-tooltip>
+    `
+
+  const hoverTooltip = createTooltip('hover')
+
+  it.each(['mouseenter', 'focus'])(
```
Comment (luiz-ortega) [inline packages/core/src/components/tooltip/tooltip.spec.ts:13]: praise: massa demais isso aqui! Não conhecia o `it.each` :rocket:

**Review by luiz-ortega — APPROVED** [review]

File: packages/core/src/components/tooltip/stories/tooltip.args.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import { Category } from '@atomium/storybook-utils/enums/table'
+
+export const TooltipStoryArgs = {
+  parameters: {
+    actions: {
+      handles: ['atomClick'],
+    },
+    docs: {
+      description: {
+        component: 'Atom-Tooltip',
```
Comment (felipefialho) [inline packages/core/src/components/tooltip/stories/tooltip.args.ts:10]: (block) we need to add some description here, you could use another component as example

File: packages/core/src/components/tooltip/tooltip.scss [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  &[data-popper-placement^='bottom'] > &__arrow {
+    top: -3px;
+  }
+
+  &[data-popper-placement^='left'] > &__arrow {
+    right: -3px;
+  }
+
+  &[data-popper-placement^='right'] > &__arrow {
+    left: -3px;
```
Comment (felipefialho) [inline packages/core/src/components/tooltip/tooltip.scss:53]: (suggestion) maybe we could transform `-3px` in a variable?

File: packages/core/src/components/tooltip/tooltip.scss [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+  &__content {
+    align-items: center;
+    display: flex;
+
+    button {
+      display: none;
+    }
+
+    @media (max-width: 768px) {
```
Comment (felipefialho) [inline packages/core/src/components/tooltip/tooltip.scss:64]: (block) we need to use `rupture` for it

```suggestion
    @include below(mobile) {
```

File: packages/core/src/components/tooltip/tooltip.scss [hunk 5]
```diff
@@ -0,0 +1,79 @@
+.atom-tooltip {
```
Comment (felipefialho) [inline packages/core/src/components/tooltip/tooltip.scss:1]: (block) is important add `scss-utils` in beginning of file

```suggestion
@import '~@atomium/scss-utils/index';

.atom-tooltip {
```

File: packages/core/src/components/tooltip/tooltip.scss [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+  &[data-popper-placement^='right'] > &__arrow {
+    left: -3px;
+  }
+
+  &__content {
+    align-items: center;
+    display: flex;
+
+    button {
```
Comment (felipefialho) [inline packages/core/src/components/tooltip/tooltip.scss:60]: (suggestion) maybe we could use a class instead modify the element directly?

File: packages/core/src/components/tooltip/tooltip.tsx
Comment (felipefialho) [inline packages/core/src/components/tooltip/tooltip.tsx:1]: (suggestion) maybe we could remove all the comments explaining the code

File: packages/core/src/components/tooltip/tooltip.tsx [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+  disconnectedCallback() {
+    this.untachEvents()
+
+    this._popperInstance.destroy()
+  }
+
+  private isMobile() {
+    return window.matchMedia('(max-width: 768px)').matches
+  }
```
Comment (felipefialho) [inline packages/core/src/components/tooltip/tooltip.tsx:115]: (question) do you test it inside vue project?

**Review by felipefialho — CHANGES_REQUESTED** [review]

File: packages/core/src/components/tooltip/tooltip.tsx
Comment (igorwessel) [reply packages/core/src/components/tooltip/tooltip.tsx:1]: It's not necessarily explaining the code, it's more for public documentation when using the component. 
Do you think it is necessary?

Captura de Tela 2024-01-12 às 15 46 28

Comment (felipefialho) [reply packages/core/src/components/tooltip/tooltip.tsx:1]: Maybe just add a short description like `Tooltip is a small pop-up box that appears when a user moves a mouse over an element`, just to keep to pattern

What u think?

File: packages/core/src/components/tooltip/tooltip.tsx (hunk 7 above)
Comment (igorwessel) [reply packages/core/src/components/tooltip/tooltip.tsx:115]: Yeah! it worked perfectly, better than react.
I ended up running into an issue that I'm investigating with SSR.

Comment (igorwessel) [reply packages/core/src/components/tooltip/tooltip.tsx:115]: https://github.com/juntossomosmais/atomium/assets/54173994/2fd68551-e663-431c-b5f9-2bf06aa1520d

File: packages/core/src/components/tooltip/tooltip.scss (hunk 4 above)
Comment (igorwessel) [reply packages/core/src/components/tooltip/tooltip.scss]: I tried using it but it didn't work, and I thought the utils were automatically imported. 
Importing resolved the problem. [46693eb](https://github.com/juntossomosmais/atomium/pull/367/commits/46693eba961f2aa486f22778890da1f7563b8950)

File: packages/core/src/components/tooltip/tooltip.scss (hunk 6 above)
Comment (igorwessel) [reply packages/core/src/components/tooltip/tooltip.scss]: done in [9d8ac12](https://github.com/juntossomosmais/atomium/pull/367/commits/9d8ac1227fbae8f69ee50c22dc65a3f99d42706c)

File: packages/core/src/components/tooltip/tooltip.scss (hunk 3 above)
Comment (igorwessel) [reply packages/core/src/components/tooltip/tooltip.scss]: I started thinking more with SCSS, and choose to use a loop to perform this section. done in [99079d1](https://github.com/juntossomosmais/atomium/pull/367/commits/99079d1bec5252d85cacc605b49d0ca4727d17ee)

File: packages/core/src/components/tooltip/stories/tooltip.args.ts (hunk 2 above)
Comment (igorwessel) [reply packages/core/src/components/tooltip/stories/tooltip.args.ts]: done in [a9fe3bb](https://github.com/juntossomosmais/atomium/pull/367/commits/a9fe3bb09787d994d0fa8dd893ac689c1a1ae128)

**Review by gabrielduete — DISMISSED** [review]

**Review by luiz-ortega — DISMISSED** [review]

**Review by gabrielduete — DISMISSED** [review]

File: packages/core/src/components/tooltip/tooltip.tsx (hunk 7 above)
Comment (felipefialho) [reply packages/core/src/components/tooltip/tooltip.tsx:120]: @MayaraRMA Is a good idea your validation here too

File: packages/core/src/components/tooltip/tooltip.tsx
Comment (felipefialho) [reply packages/core/src/components/tooltip/tooltip.tsx:1]: @igorwessel Just a remember hahaha

Comment (igorwessel) [reply packages/core/src/components/tooltip/tooltip.tsx:1]: >Maybe just add a short description like Tooltip is a small pop-up box that appears when a user moves a mouse over an element, just to keep to pattern

I added this documentation to storybook and removed the code comments.

File: packages/core/src/components/tooltip/stories/tooltip.react.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      <div>{args.text}</div>
+    </AtomTooltip>
+  </>
+)
+
+export const Default: StoryObj = {
+  render: (args) => createTooltip(args),
+  args: {
+    ...TooltipComponentArgs,
+  },
```
Comment (MayaraRMA) [inline packages/core/src/components/tooltip/stories/tooltip.react.stories.tsx:37]: @igorwessel no preview o default não está funcionando

**Review by MayaraRMA — CHANGES_REQUESTED** [review]

File: packages/core/src/components/tooltip/stories/tooltip.react.stories.tsx (hunk 8 above)
Comment (MayaraRMA) [reply packages/core/src/components/tooltip/stories/tooltip.react.stories.tsx:37]: https://github.com/juntossomosmais/atomium/assets/37077860/230e9d94-fad9-4b29-806b-60b83eb1f66c

File: package.json [hunk 9]
```diff
@@ -30,6 +30,7 @@
   },
   "dependencies": {
     "@ionic/core": "^7.6.2",
+    "@popperjs/core": "^2.11.8",
```
Comment (MayaraRMA) [inline package.json:33]: já existe o componente `ion-popover` era necessário usar essa biblioteca?

File: packages/core/src/components/tooltip/stories/tooltip.react.stories.tsx (hunk 8 above)
Comment (igorwessel) [reply packages/core/src/components/tooltip/stories/tooltip.react.stories.tsx:37]: Descobri o problema mas não sei se possui uma solução, no "default" não funciona por que tem dois elementos com o mesmo ID para se ancorar e nisso ele se perde. Não sei dizer se tem como desabilitar o story que é criado para a documentação pelo menos não encontrei nada no storybook.

File: package.json (hunk 9 above)
Comment (igorwessel) [reply package.json:33]: Foi necessário pela complexidade de alterar o ion-popover para se adequar as necessidades do tooltip, envolvia muitas alterações no componente ion-popover para se comportar como tooltip e essas alterações envolvia algumas coisas como: ter que retirar uma classe que faz o lock do scroll no site quando usado o ion-popover, adicionar o acompanhamento para o scroll e responsividade no quesito de automaticamente mudar de lado/direção.

**Comment by igorwessel** [comment]
@MayaraRMA @felipefialho I will create the ADR on why it was necessary to use a library.

**Review by felipefialho — APPROVED** [review]

**Review by gabrielduete — APPROVED** [review]
:dev-ninja:

File: apps/docs/stories/architecture-decision-records/0014-why-use-popperjs.stories.mdx [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+These changes included things like:
+
+- Removing a class that locks the scroll on the site (added to the body) when using the `ion-popover`,
+- Adding scroll tracking
+- Ensuring responsiveness in terms of automatically changing side/direction.
+
+## Decision Drivers
+
+- Bundle Size: The size of the library was considered. If the size was too large, it would be a NO-GO for inclusion in the project. As the entire library currently (2024/11/05) only 8.9KB, we understood that the size would not negatively impact the user experience.
```
[the remark this answers is not in the document]
Comment (igorwessel) [reply apps/docs/stories/architecture-decision-records/0014-why-use-popperjs.stories.mdx]: done in [4b6bf6c](https://github.com/juntossomosmais/atomium/pull/367/commits/4b6bf6cd46b8ac5555152893dc83c62623f3e446)

File: apps/docs/stories/architecture-decision-records/0014-why-use-popperjs.stories.mdx [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+## Context
+
+It was necessary due to the complexity of modifying the `ion-popover` to meet the tooltip requirements. This involved many changes to the `ion-popover` component to make it behave like a tooltip.
+
+## Problems
+
+These changes included things like:
+
+- Removing a class that locks the scroll on the site (added to the body) when using the `ion-popover`,
+- Adding scroll tracking
```
[the remark this answers is not in the document]
Comment (igorwessel) [reply apps/docs/stories/architecture-decision-records/0014-why-use-popperjs.stories.mdx]: done in [4b6bf6c](https://github.com/juntossomosmais/atomium/pull/367/commits/4b6bf6cd46b8ac5555152893dc83c62623f3e446)

File: apps/docs/stories/architecture-decision-records/0014-why-use-popperjs.stories.mdx [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+## Context
+
+It was necessary due to the complexity of modifying the `ion-popover` to meet the tooltip requirements. This involved many changes to the `ion-popover` component to make it behave like a tooltip.
+
+## Problems
+
+These changes included things like:
+
+- Removing a class that locks the scroll on the site (added to the body) when using the `ion-popover`,
```
[the remark this answers is not in the document]
Comment (igorwessel) [reply apps/docs/stories/architecture-decision-records/0014-why-use-popperjs.stories.mdx]: done in [4b6bf6c](https://github.com/juntossomosmais/atomium/pull/367/commits/4b6bf6cd46b8ac5555152893dc83c62623f3e446)

**Review by gabrielduete — APPROVED** [review]

**Review by felipefialho — APPROVED** [review]

**Review by felipefialho — APPROVED** [review]

**Review by gabrielduete — APPROVED** [review]

**Review by luiz-ortega — APPROVED** [review]

File: packages/core/src/components/tooltip/tooltip.tsx (hunk 7 above)
Comment (igorwessel) [reply packages/core/src/components/tooltip/tooltip.tsx:145]: @MayaraRMA @felipefialho you able to validate?

**Review by MayaraRMA — APPROVED** [review]

**Pull request merged**